File: backend/src/test/java/dev/coldhands/pair/stairs/backend/WithBackendHttpClient.java
package dev.coldhands.pair.stairs.backend;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import dev.coldhands.pair.stairs.backend.domain.*;
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.CalculateInputDto;
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.SaveCombinationEventDto;
import org.springframework.http.*;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.web.client.RestTemplate;

import java.time.LocalDate;
import java.util.List;

import static java.time.format.DateTimeFormatter.ISO_DATE;
import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.http.HttpMethod.DELETE;

public interface WithBackendHttpClient {

    String BASE_URL = "http://localhost:8080";
    String IDP_TOKEN_URL = "http://localhost:5556/dex/token";
    String OAUTH_CLIENT_ID = "pair-stairs";
    String OAUTH_CLIENT_SECRET = "ZXhhbXBsZS1hcHAtc2VjcmV0";
    String USERNAME = "admin@example.com";
    String PASSWORD = "password";

    ObjectMapper OBJECT_MAPPER = new ObjectMapper()
            .registerModule(new JavaTimeModule());

    RestTemplate REST_TEMPLATE = initialiseRestTemplate();

    private static RestTemplate initialiseRestTemplate() {
        final var restTemplate = new RestTemplate();
        final String jwt = fetchJwt(restTemplate);

        restTemplate.getInterceptors().add(((request, body, execution) -> {
            request.getHeaders().setBearerAuth(jwt);
            return execution.execute(request, body);
        }));

        return restTemplate;
    }

    default long createDeveloper(String developerName) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<String> request = new HttpEntity<>("""
                {
                  "name": "%s"
                }""".formatted(developerName), headers);

        ResponseEntity<Developer> response = REST_TEMPLATE.postForEntity(STR."\{BASE_URL}/api/v1/developers", request, Developer.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);

        final Developer developer = response.getBody();
        return developer.id();
    }

    default List<Long> getDeveloperIdsFor(List<String> developerNames) throws Exception {
        final ResponseEntity<String> response = REST_TEMPLATE.getForEntity(STR."\{BASE_URL}/api/v1/developers", String.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);

        final String responseBody = response.getBody();

        final List<Developer> developers = OBJECT_MAPPER.readValue(responseBody, new TypeReference<>() {
        });

        return developers.stream()
                .filter(developer -> developerNames.contains(developer.name()))
                .map(Developer::id)
                .toList();
    }

    default long createStream(String streamName) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<String> request = new HttpEntity<>("""
                {
                  "name": "%s"
                }""".formatted(streamName), headers);

        ResponseEntity<Stream> response = REST_TEMPLATE.postForEntity(STR."\{BASE_URL}/api/v1/streams", request, Stream.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);

        final Stream stream = response.getBody();
        return stream.id();
    }

    default List<Long> getStreamIdsFor(List<String> streamNames) throws Exception {
        final ResponseEntity<String> response = REST_TEMPLATE.getForEntity(STR."\{BASE_URL}/api/v1/streams", String.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);

        final String responseBody = response.getBody();

        final List<Stream> streams = OBJECT_MAPPER.readValue(responseBody, new TypeReference<>() {
        });

        return streams.stream()
                .filter(stream -> streamNames.contains(stream.name()))
                .map(Stream::id)
                .toList();
    }

    default List<ScoredCombination> calculateCombinations(List<Long> developerIds, List<Long> streamIds) throws Exception {
        final CalculateInputDto input = new CalculateInputDto(developerIds, streamIds);

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<String> request = new HttpEntity<>(OBJECT_MAPPER.writeValueAsString(input), headers);

        ResponseEntity<String> response = REST_TEMPLATE.postForEntity(STR."\{BASE_URL}/api/v1/combinations/calculate", request, String.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);

        final String responseBody = response.getBody();

        return OBJECT_MAPPER.readValue(responseBody, new TypeReference<>() {
        });
    }

    default void saveCombinationEventFor(LocalDate date, List<PairStream> combination) throws Exception {
        List<SaveCombinationEventDto.PairStreamByIds> combinationByIds = combination.stream()
                .map(ps -> {
                    List<Long> developerIds = ps.developers().stream()
                            .map(DeveloperInfo::id)
                            .toList();
                    return new SaveCombinationEventDto.PairStreamByIds(developerIds, ps.stream().id());
                })
                .toList();

        final SaveCombinationEventDto dto = new SaveCombinationEventDto(date, combinationByIds);

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<String> request = new HttpEntity<>(OBJECT_MAPPER.writeValueAsString(dto), headers);

        ResponseEntity<Void> response = REST_TEMPLATE.postForEntity(STR."\{BASE_URL}/api/v1/combinations/event", request, Void.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);
    }

    default List<CombinationEvent> getCombinationEvents() throws Exception {
        ResponseEntity<String> response = REST_TEMPLATE.getForEntity(STR."\{BASE_URL}/api/v1/combinations/event", String.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);

        final String responseBody = response.getBody();

        return OBJECT_MAPPER.readValue(responseBody, new TypeReference<>() {
        });
    }

    default void deleteCombinationEvent(long id) {
        ResponseEntity<Void> response = REST_TEMPLATE.exchange(STR."\{BASE_URL}/api/v1/combinations/event/{id}", DELETE, null, Void.class, id);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NO_CONTENT);
    }

    default DeveloperStats getDeveloperStatsBetween(long developerId, LocalDate startDate, LocalDate endDate) throws Exception {
        final ResponseEntity<String> response = REST_TEMPLATE.getForEntity(STR."\{BASE_URL}/api/v1/developers/\{developerId}/stats?startDate=\{startDate.format(ISO_DATE)}&endDate=\{endDate.format(ISO_DATE)}", String.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);

        final String responseBody = response.getBody();

        return OBJECT_MAPPER.readValue(responseBody, new TypeReference<>() {
        });
    }

    default StreamStats getStreamStatsBetween(long streamId, LocalDate startDate, LocalDate endDate) throws Exception {
        final ResponseEntity<String> response = REST_TEMPLATE.getForEntity(STR."\{BASE_URL}/api/v1/streams/\{streamId}/stats?startDate=\{startDate.format(ISO_DATE)}&endDate=\{endDate.format(ISO_DATE)}", String.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);

        final String responseBody = response.getBody();

        return OBJECT_MAPPER.readValue(responseBody, new TypeReference<>() {
        });
    }

    record TokenResponse(String access_token) {
    }

    private static String fetchJwt(RestTemplate restTemplate) {
        var form = new LinkedMultiValueMap<String, String>();
        form.add("grant_type", "password");
        form.add("username", USERNAME);
        form.add("password", PASSWORD);
        form.add("scope", "openid profile email"); // add offline_access if needed

        var headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
        headers.setBasicAuth(OAUTH_CLIENT_ID, OAUTH_CLIENT_SECRET);

        var response = restTemplate.postForEntity(IDP_TOKEN_URL, new HttpEntity<>(form, headers), TokenResponse.class);

        return response.getBody().access_token();
    }

}

File: backend/src/test/kotlin/dev/coldhands/pair/stairs/backend/WithBackendHttpClient.kt
package dev.coldhands.pair.stairs.backend

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import dev.coldhands.pair.stairs.backend.domain.*
import dev.coldhands.pair.stairs.backend.domain.developer.Developer
import dev.coldhands.pair.stairs.backend.domain.developer.DeveloperStats
import dev.coldhands.pair.stairs.backend.domain.stream.Stream
import dev.coldhands.pair.stairs.backend.domain.stream.StreamStats
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.CalculateInputDto
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.SaveCombinationEventDto
import io.kotest.matchers.shouldBe
import org.springframework.http.*
import org.springframework.http.HttpMethod.DELETE
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.client.RestTemplate
import java.time.LocalDate
import java.time.format.DateTimeFormatter.ISO_DATE

interface WithBackendHttpClient {

    fun createDeveloper(developerName: String): DeveloperId {
        val headers = HttpHeaders().apply { contentType = MediaType.APPLICATION_JSON }
        val payload = """
            {
              "name": "$developerName"
            }
        """.trimIndent()
        val request = HttpEntity(payload, headers)

        val response: ResponseEntity<Developer> =
            REST_TEMPLATE.postForEntity("$BASE_URL/api/v1/developers", request, Developer::class.java)

        response.statusCode shouldBe HttpStatus.CREATED
        return response.body!!.id
    }

    fun getDeveloperIdsFor(developerNames: List<String>): List<DeveloperId> {
        val response: ResponseEntity<String> =
            REST_TEMPLATE.getForEntity("$BASE_URL/api/v1/developers", String::class.java)

        response.statusCode shouldBe HttpStatus.OK

        val developers: List<Developer> = OBJECT_MAPPER.readValue(response.body!!)
        return developers
            .filter { it.name in developerNames }
            .map(Developer::id)
    }

    fun createStream(streamName: String): StreamId {
        val headers = HttpHeaders().apply { contentType = MediaType.APPLICATION_JSON }
        val payload = """
            {
              "name": "$streamName"
            }
        """.trimIndent()
        val request = HttpEntity(payload, headers)

        val response: ResponseEntity<Stream> =
            REST_TEMPLATE.postForEntity("$BASE_URL/api/v1/streams", request, Stream::class.java)

        response.statusCode shouldBe HttpStatus.CREATED
        return response.body!!.id
    }

    fun getStreamIdsFor(streamNames: List<String>): List<StreamId> {
        val response: ResponseEntity<String> =
            REST_TEMPLATE.getForEntity("$BASE_URL/api/v1/streams", String::class.java)

        response.statusCode shouldBe HttpStatus.OK

        val streams: List<Stream> = OBJECT_MAPPER.readValue(response.body!!)
        return streams
            .filter { it.name in streamNames }
            .map(Stream::id)
    }

    fun calculateCombinations(developerIds: List<DeveloperId>, streamIds: List<StreamId>): List<ScoredCombination> {
        val input = CalculateInputDto(developerIds, streamIds)

        val headers = HttpHeaders().apply { contentType = MediaType.APPLICATION_JSON }
        val request = HttpEntity(OBJECT_MAPPER.writeValueAsString(input), headers)

        val response: ResponseEntity<String> =
            REST_TEMPLATE.postForEntity("$BASE_URL/api/v1/combinations/calculate", request, String::class.java)

        response.statusCode shouldBe HttpStatus.OK

        return OBJECT_MAPPER.readValue(response.body!!)
    }

    fun saveCombinationEventFor(date: LocalDate, combination: List<PairStream>) {
        val combinationByIds: List<SaveCombinationEventDto.PairStreamByIds> =
            combination.map { ps ->
                val developerIds = ps.developers().map(DeveloperInfo::id)
                // todo just use DeveloperId directly once DeveloperInfo is kotlin-ed
                SaveCombinationEventDto.PairStreamByIds(developerIds.map { DeveloperId(it) }, StreamId(ps.stream().id()))
            }

        val dto = SaveCombinationEventDto(date, combinationByIds)

        val headers = HttpHeaders().apply { contentType = MediaType.APPLICATION_JSON }
        val request = HttpEntity(OBJECT_MAPPER.writeValueAsString(dto), headers)

        val response: ResponseEntity<Void> =
            REST_TEMPLATE.postForEntity("$BASE_URL/api/v1/combinations/event", request, Void::class.java)

        response.statusCode shouldBe HttpStatus.CREATED
    }

    fun getCombinationEvents(): List<CombinationEvent> {
        val response: ResponseEntity<String> =
            REST_TEMPLATE.getForEntity("$BASE_URL/api/v1/combinations/event", String::class.java)

        response.statusCode shouldBe HttpStatus.OK

        return OBJECT_MAPPER.readValue(response.body!!)
    }

    fun deleteCombinationEvent(id: Long) {
        val response: ResponseEntity<Void> =
            REST_TEMPLATE.exchange("$BASE_URL/api/v1/combinations/event/{id}", DELETE, null, Void::class.java, id)

        response.statusCode shouldBe HttpStatus.NO_CONTENT
    }

    fun getDeveloperStatsBetween(developerId: DeveloperId, startDate: LocalDate, endDate: LocalDate): DeveloperStats {
        val url =
            "$BASE_URL/api/v1/developers/${developerId.value}/stats?startDate=${startDate.format(ISO_DATE)}&endDate=${endDate.format(ISO_DATE)}"

        val response: ResponseEntity<String> = REST_TEMPLATE.getForEntity(url, String::class.java)

        response.statusCode shouldBe HttpStatus.OK

        return OBJECT_MAPPER.readValue(response.body!!)
    }

    fun getStreamStatsBetween(streamId: StreamId, startDate: LocalDate, endDate: LocalDate): StreamStats {
        val url =
            "$BASE_URL/api/v1/streams/${streamId.value}/stats?startDate=${startDate.format(ISO_DATE)}&endDate=${endDate.format(ISO_DATE)}"

        val response: ResponseEntity<String> = REST_TEMPLATE.getForEntity(url, String::class.java)

        response.statusCode shouldBe HttpStatus.OK

        return OBJECT_MAPPER.readValue(response.body!!)
    }

    data class TokenResponse(
        @JsonProperty("access_token") val accessToken: String)

    companion object {
        const val BASE_URL: String = "http://localhost:8080"
        const val IDP_TOKEN_URL: String = "http://localhost:5556/dex/token"
        const val OAUTH_CLIENT_ID: String = "pair-stairs"
        const val OAUTH_CLIENT_SECRET: String = "ZXhhbXBsZS1hcHAtc2VjcmV0"
        const val USERNAME: String = "admin@example.com"
        const val PASSWORD: String = "password"

        private val OBJECT_MAPPER: ObjectMapper = jacksonObjectMapper()
            .registerModule(JavaTimeModule())

        val REST_TEMPLATE: RestTemplate = initialiseRestTemplate()

        private fun initialiseRestTemplate(): RestTemplate {
            val rt = RestTemplate()
            val jwt = fetchJwt(rt)
            rt.interceptors.add { request, body, execution ->
                request.headers.setBearerAuth(jwt)
                execution.execute(request, body)
            }
            return rt
        }

        private fun fetchJwt(restTemplate: RestTemplate): String {
            val form = LinkedMultiValueMap<String, String>().apply {
                add("grant_type", "password")
                add("username", USERNAME)
                add("password", PASSWORD)
                add("scope", "openid profile email") // add offline_access if needed
            }

            val headers = HttpHeaders().apply {
                contentType = MediaType.APPLICATION_FORM_URLENCODED
                setBasicAuth(OAUTH_CLIENT_ID, OAUTH_CLIENT_SECRET)
            }

            val response = restTemplate.postForEntity(
                IDP_TOKEN_URL,
                HttpEntity(form, headers),
                TokenResponse::class.java
            )

            return response.body!!.accessToken
        }
    }
}
